File: datasets/Part 2 - Regression/Section 5 - Multiple Linear Regression/1_Regresion_Lineal_Multiple.py
# -*- coding: utf-8 -*-
"""
Created on Thu Feb  2 10:51:35 2023

@author: Imanol
"""
# Regresión lineal Multiple

# Cómo importar las librerías
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


# Importar el data set
dataset = pd.read_csv("50_Startups.csv")
X = dataset.iloc[:,:-1].values
y = dataset.iloc[:,-1].values

# Codificar datos categóricos
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
ct = ColumnTransformer(
    [('one_hot_encoder', OneHotEncoder(categories='auto'), [-1])],
    remainder='passthrough')
X = np.array(ct.fit_transform(X), dtype=np.float)

# Evitar la trampa de las variables ficticias
# Eliminamos una de las columnas Dummy 
X = X[:,1:]

# Dividir el data set en training y testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0) 

# Ajustar el modulo de Regresión lineal múltiple con el conjunto de entrenamiento
from sklearn.linear_model import LinearRegression
regression = LinearRegression()
regression.fit(X_train, y_train)

# Predicción de los resultados en el conjunto de testing
y_pred = regression.predict(X_test)

# Contruir el modelo óptimo de RLM utilizando la Eliminación hacia atrás
import statsmodels.api as sm
X = np.append(arr = np.ones((50,1)).astype(int), values=X, axis=1)
SL = 0.05
## Creamos una regresión con todas las variables utilizando la clase statmodels que nos devuelve información util
X_opt = X[:, [0, 1, 2, 3, 4, 5]].tolist()
regression_OLS = sm.OLS(endog= y, exog= X_opt).fit()
regression_OLS.summary()
## Eliminamos columna numero 2
X_opt = X[:, [0, 1, 3, 4, 5]].tolist()
regression_OLS = sm.OLS(endog= y, exog= X_opt).fit()
regression_OLS.summary()
## Eliminamos columna numero 1
X_opt = X[:, [0, 3, 4, 5]].tolist()
regression_OLS = sm.OLS(endog= y, exog= X_opt).fit()
regression_OLS.summary()
## Eliminamos columna numero 4
X_opt = X[:, [0, 3, 5]].tolist()
regression_OLS = sm.OLS(endog= y, exog= X_opt).fit()
regression_OLS.summary()
## Eliminamos columna numero 5
X_opt = X[:, [0, 3]].tolist()
regression_OLS = sm.OLS(endog= y, exog= X_opt).fit()
regression_OLS.summary()

# Automatizandolo
""" 
import statsmodels.api as sm
def backwardElimination(x, sl):    
   numVars = len(x[0])    
   for i in range(0, numVars):        
     regressor_OLS = sm.OLS(y, x.tolist()).fit()        
     maxVar = max(regressor_OLS.pvalues).astype(float)        
     if maxVar > sl:            
        for j in range(0, numVars - i):                
          if (regressor_OLS.pvalues[j].astype(float) == maxVar):                    
             x = np.delete(x, j, 1)    
   print(regressor_OLS.summary())  
   return x

SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_Modeled = backwardElimination(X_opt, SL)
"""

# Automatizandolo en función de p-valor y R cuadrada ajustada
"""
import statsmodels.api as sm
def backwardElimination(x, SL):
    numVars = len(x[0])    
    temp = np.zeros((50,6)).astype(int)
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x.tolist()).fit()        
        maxVar = max(regressor_OLS.pvalues).astype(float)        
        adjR_before = regressor_OLS.rsquared_adj.astype(float)        
        if maxVar > SL:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    temp[:,j] = x[:, j]                    
                    x = np.delete(x, j, 1)                    
                    tmp_regressor = sm.OLS(y, x.tolist()).fit()                    
                    adjR_after = tmp_regressor.rsquared_adj.astype(float)
                    
                    if (adjR_before >= adjR_after):
                        x_rollback = np.hstack((x, temp[:,[0,j]]))                        
                        x_rollback = np.delete(x_rollback, j, 1)    
                        print (regressor_OLS.summary())                        
                        return x_rollback 
                    else:
                        continue
    regressor_OLS.summary()
    return x            
        

SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_Modeled = backwardElimination(X_opt, SL)
"""

File: datasets/Part 2 - Regression/Section 5 - Multiple Linear Regression/1_Regresion_Lineal_Multiple.R
# Regresión Lineal Multiple

# Importar el dataset
dataset <- read.csv('50_Startups.csv')

# Codificar las variables categ?ricas
dataset$State <- factor(dataset$State,
                          levels = c("New York", "California", "Florida"),
                          labels = c(1, 2, 3))


# Dividir los datos en training y testing
#install.packages("caTools")
library(caTools)
set.seed(123)
split <- sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

# Ajustar el modelo de regresión lineal múltiple con el conjunto de entrenamiento
regression = lm(formula = Profit ~ ., 
               data = training_set)
summary(regression)

# Predecir resultados con el conjunto de Testing
y_pred = predict(regression, newdata = testing_set)

# Construir un modelo óptimo con la Eliminación hacia atrás
SL = 0.05
regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State, 
               data = training_set)
summary(regression)
# Eliminamos State
regression = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend, 
                data = training_set)
summary(regression)
# Eliminamos Administration
regression = lm(formula = Profit ~ R.D.Spend + Marketing.Spend, 
                data = training_set)
summary(regression)
# Eliminamos Marketing.Spend
regression = lm(formula = Profit ~ R.D.Spend, 
                data = training_set)
summary(regression)

# Automatización 
## Función
backwardElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~ ., data = x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[, -j]
    }
    numVars = numVars - 1
  }
  return(summary(regressor))
}
## Aplicación de la función
SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(training_set, SL)